Arbitrary Gyro and Sensor alignment

hydra opened the pull request:
Allow arbitrary gyro and compass alignment.

* Rotation matrix is pre-computed, after sensor detection completes.
* Supports dual gyro boards with offsets other than 90 degrees.
* Supports external compasses mounted at any angle.
* MSP support included, backwards compatible where possible.
* Alignment values are now OFFSETS from target/sensor defaults, previous
alignment values overrode driver defaults.
* Required for SPRacingH7EXTREME - http://seriouslypro.com/spracingh7extreme

MSP clients can detect arbitrary alignment by looking at the
'multi-gyro' value which now uses another bit to indicate arbitrary
alignment, also the size of the SENSOR_ALIGNMENT is now 18 bytes longer.

* Initial testing done on SPRacingF7DUAL.
* Flight tested on the SPRacingH7EXTREME which has second gyro at 45 degrees.
* No noticeable change to PID loop speed, see results below.

before:
```
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:       15528 B        16 KB     94.78%
      ITCM_FLASH:          0 GB        16 KB      0.00%
ITCM_FLASH_CONFIG:          0 GB        16 KB      0.00%
     ITCM_FLASH1:          0 GB       480 KB      0.00%
      AXIM_FLASH:        8274 B        16 KB     50.50%
AXIM_FLASH_CONFIG:          0 GB        16 KB      0.00%
     AXIM_FLASH1:      388968 B       480 KB     79.14%
        DTCM_RAM:       24920 B        64 KB     38.02%
           SRAM1:       42744 B       176 KB     23.72%
           SRAM2:          0 GB        16 KB      0.00%
       MEMORY_B1:          0 GB         0 GB     -1.#J%
   text	   data	    bss	    dec	    hex	filename
 391246	   5996	  61668	 458910	  7009e	./obj/main/betaflight_SPRACINGF7DUAL.elf

Task list             rate/hz  max/us  avg/us maxload avgload     total/ms
00 - (         SYSTEM)      9       2       1    0.5%    0.5%         0
01 - (         SYSTEM)    990     324       2   32.5%    0.6%       495
02 - (            PID)   1982      69      32   14.1%    6.8%     13433
   - (           GYRO)   7931
03 - (            ACC)    990      21       9    2.5%    1.3%       466
04 - (       ATTITUDE)     99      17      10    0.6%    0.5%        56
05 - (             RX)     32      50      34    0.6%    0.6%        66
06 - (         SERIAL)     99     128       2    1.7%    0.5%       143
08 - (BATTERY_VOLTAGE)     49       4       2    0.5%    0.5%         5
09 - (BATTERY_CURRENT)     49       3       1    0.5%    0.5%         2
10 - ( BATTERY_ALERTS)      5       3       2    0.5%    0.5%         0
11 - (         BEEPER)     99       6       1    0.5%    0.5%         6
14 - (           BARO)     43     112      67    0.9%    0.7%       163
15 - (       ALTITUDE)     39       9       6    0.5%    0.5%        11
17 - (      TELEMETRY)    249      80       0    2.4%    0.0%       896
18 - (       LEDSTRIP)     99      64       5    1.1%    0.5%        35
19 - (    TRANSPONDER)    248      10       1    0.7%    0.5%         9
20 - (            OSD)     59     477      13    3.3%    0.5%        43
22 - (            CMS)     59       2       1    0.5%    0.5%         3
23 - (        VTXCTRL)      5  401006       7  201.0%    0.5%       504
24 - (        CAMCTRL)      4       1       1    0.5%    0.5%         0
26 - (    ADCINTERNAL)      1       4       3    0.5%    0.5%         0
27 - (       PINIOBOX)     19       2       1    0.5%    0.5%         0
RX Check Function                   9       1                         1
Total (excluding SERIAL)                       264.7%   17.5%
```

after:

```
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:       15448 B        16 KB     94.29%
      ITCM_FLASH:          0 GB        16 KB      0.00%
ITCM_FLASH_CONFIG:          0 GB        16 KB      0.00%
     ITCM_FLASH1:          0 GB       480 KB      0.00%
      AXIM_FLASH:        8274 B        16 KB     50.50%
AXIM_FLASH_CONFIG:          0 GB        16 KB      0.00%
     AXIM_FLASH1:      389632 B       480 KB     79.27%
        DTCM_RAM:       25028 B        64 KB     38.19%
           SRAM1:       42816 B       176 KB     23.76%
           SRAM2:          0 GB        16 KB      0.00%
       MEMORY_B1:          0 GB         0 GB     -1.#J%
   text	   data	    bss	    dec	    hex	filename
 391910	   5996	  61848	 459754	  703ea	./obj/main/betaflight_SPRACINGF7DUAL.elf

Task list             rate/hz  max/us  avg/us maxload avgload     total/ms
00 - (         SYSTEM)      9       1       0    0.5%    0.0%         0
01 - (         SYSTEM)    991     325       1   32.7%    0.5%       436
02 - (            PID)   1980      69      32   14.1%    6.8%      4597
   - (           GYRO)   7921
03 - (            ACC)    991      17       8    2.1%    1.2%       164
04 - (       ATTITUDE)     99      12      10    0.6%    0.5%        19
05 - (             RX)     32      50      35    0.6%    0.6%        23
06 - (         SERIAL)     99     367       2    4.1%    0.5%        31
08 - (BATTERY_VOLTAGE)     49       4       2    0.5%    0.5%         2
09 - (BATTERY_CURRENT)     49       1       1    0.5%    0.5%         0
10 - ( BATTERY_ALERTS)      5       3       2    0.5%    0.5%         0
11 - (         BEEPER)     99       6       1    0.5%    0.5%         2
14 - (           BARO)     43     103      69    0.9%    0.7%        59
15 - (       ALTITUDE)     39       7       5    0.5%    0.5%         4
17 - (      TELEMETRY)    248      81       1    2.5%    0.5%       218
18 - (       LEDSTRIP)     99      63       8    1.1%    0.5%        13
19 - (    TRANSPONDER)    248       2       1    0.5%    0.5%         3
20 - (            OSD)     59     477      12    3.3%    0.5%        16
22 - (            CMS)     59       2       1    0.5%    0.5%         1
23 - (        VTXCTRL)      5  401007     643  201.0%    0.8%       503
24 - (        CAMCTRL)      5       1       1    0.5%    0.5%         0
26 - (    ADCINTERNAL)      1       4       1    0.5%    0.5%         0
27 - (       PINIOBOX)     19       2       1    0.5%    0.5%         0
RX Check Function                   3       1                         0
Total (excluding SERIAL)                       264.4%   17.6%
```

mikeller commented on src/main/msp/msp_protocol.h:
```diff
@@ -62,7 +62,7 @@
 
 #define API_VERSION_MAJOR                   1  // increment when major changes are made
-#define API_VERSION_MINOR                   41 // increment after a release, to set the version for all changes to go into the following release (if no changes to MSP are made between the releases, this can be reverted before the release)
+#define API_VERSION_MINOR                   42 // increment after a release, to set the version for all changes to go into the following release (if no changes to MSP are made between the releases, this can be reverted before the release)
```
This should not be in here.

hydra replied:
yes I understand that, however the comments in the MSP code won't match the code.

I targeted this for 4.1, so when 4.0 is done the API version should be bumped to 42 by another commit and this change will disappear right?  No problem removing for you now if you like though.

mikeller replied:
Yes please, remove. Otherwise we'll get back to the bad old days where every pull request including an MSP change will increment this 'because it's needed', leading to an increment of 10 or so between releases...

hydra replied:
done

mikeller commented on src/test/Makefile:
```diff
@@ -394,10 +395,22 @@ COMMON_FLAGS = \
 
 ifeq ($(shell $(CC) -v 2>&1 | grep -q "clang version" && echo "clang"),clang)
 COMMON_FLAGS += -fblocks
+ifndef CYGWIN
```
This does not seem to be related.

hydra replied:
yes I have removed that commit.

hydra replied:
see #7846

mikeller commented on src/main/cli/settings.c:
```diff
@@ -1361,13 +1350,17 @@ const clivalue_t valueTable[] = {
     { "gyro_2_spibus",  VAR_UINT8 | HARDWARE_VALUE, .config.minmaxUnsigned = { 0, SPIDEV_COUNT }, PG_GYRO_DEVICE_CONFIG, PG_ARRAY_ELEMENT_OFFSET(gyroDeviceConfig_t, 1, spiBus) },
     { "gyro_2_i2cBus",  VAR_UINT8 | HARDWARE_VALUE, .config.minmaxUnsigned = { 0, I2CDEV_COUNT }, PG_GYRO_DEVICE_CONFIG, PG_ARRAY_ELEMENT_OFFSET(gyroDeviceConfig_t, 1, i2cBus) },
     { "gyro_2_i2c_address", VAR_UINT8  | HARDWARE_VALUE, .config.minmaxUnsigned = { 0, I2C_ADDR7_MAX }, PG_GYRO_DEVICE_CONFIG, PG_ARRAY_ELEMENT_OFFSET(gyroDeviceConfig_t, 1, i2cAddress) },
-    { "gyro_2_sensor_align", VAR_UINT8  | HARDWARE_VALUE | MODE_LOOKUP, .config.lookup = { TABLE_ALIGNMENT }, PG_GYRO_DEVICE_CONFIG, PG_ARRAY_ELEMENT_OFFSET(gyroDeviceConfig_t, 1, align) },
```
It would be nice if we could keep the old parameters for a transitional period, and convert them to new values on the fly. But not sure how involved this will be.

hydra replied:
I don't think it's possible, you'd have to maintain and convert between two sets of alignment settings + a bunch of conditional code and likely a mess should they ever conflict.  One setting keeps things simple and consistent

It's only possible to go from a legacy sensor alignment value to a set of angles, but not the other way round.

See the MSP backwards compatibility for a taste of the situation.

mikeller replied:
Good point. Will need manual updates from users then.

stawiski commented on src/main/sensors/boardalignment.c:
```diff
@@ -59,69 +61,55 @@ void initBoardAlignment(const boardAlignment_t *boardAlignment)
+
+    buildAlignmentFromRotation(&sensorAlignment, rotation);
+
+#if 0
```
Umm... you have ```#if 0``` there.

hydra replied:
will delete that now, it was temporary code during refactoring process that allowed me to keep the tests running, thanks for spotting it!

hydra commented:
This is now flight-tested on a new FC.

hydra commented:
* rebased against master
* improved backwards compatibility
* added some unit tests to prove some of the backwards compatibility logic for standard/arbitrary rotation scenarios.

hydra commented:
removed feature-cut commits, we can handle that later.

some builds are broken due to flash size issues with the support for legacy MSP clients.

hydra commented:
I created another commit here to delete the legacy MSP support which saves a around 300 bytes of flash space on F7.

https://github.com/hydra/betaflight/commit/b2e18af765d196f75d02f6b224924c4e8e1fb9bd

before:
```
Linking SPRACINGF7DUAL 
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:       15448 B        16 KB     94.29%
      ITCM_FLASH:          0 GB        16 KB      0.00%
ITCM_FLASH_CONFIG:          0 GB        16 KB      0.00%
     ITCM_FLASH1:          0 GB       480 KB      0.00%
      AXIM_FLASH:        8274 B        16 KB     50.50%
AXIM_FLASH_CONFIG:          0 GB        16 KB      0.00%
     AXIM_FLASH1:      390656 B       480 KB     79.48%
        DTCM_RAM:       25028 B        64 KB     38.19%
           SRAM1:       42816 B       176 KB     23.76%
           SRAM2:          0 GB        16 KB      0.00%
       MEMORY_B1:          0 GB         0 GB     -1.#J%
   text	   data	    bss	    dec	    hex	filename
 392934	   5996	  61848	 460778	  707ea	./obj/main/betaflight_SPRACINGF7DUAL.elf
```
after:
```
Linking SPRACINGF7DUAL 
Memory region         Used Size  Region Size  %age Used
        ITCM_RAM:       15448 B        16 KB     94.29%
      ITCM_FLASH:          0 GB        16 KB      0.00%
ITCM_FLASH_CONFIG:          0 GB        16 KB      0.00%
     ITCM_FLASH1:          0 GB       480 KB      0.00%
      AXIM_FLASH:        8274 B        16 KB     50.50%
AXIM_FLASH_CONFIG:          0 GB        16 KB      0.00%
     AXIM_FLASH1:      390368 B       480 KB     79.42%
        DTCM_RAM:       25028 B        64 KB     38.19%
           SRAM1:       42816 B       176 KB     23.76%
           SRAM2:          0 GB        16 KB      0.00%
       MEMORY_B1:          0 GB         0 GB     -1.#J%
   text	   data	    bss	    dec	    hex	filename
 392646	   5996	  61848	 460490	  706ca	./obj/main/betaflight_SPRACINGF7DUAL.elf
```

We can choose how to proceed in due course.  Comments welcome.

mikeller commented:
This will provide no value to 99% of users, while consuming more resources, and making their configuration more complicated. For this reason I think it is a better approach to make this available as a feature, and if disable provide the existing configuration that only supports 90 degree increments.

Also, this should be CLI only for 4.1, with MSP support added only for the following version.

hydra commented:
rebased against master

hydra commented:
To recap from slack conversation regarding this:

* add new feature flag
* introduce conditional code so that arbitrary gyro alignment uses the new feature flag.
* revert MSP changes
* make MSP output align_default when feature enabled.
* make MSP ignore attempts to set alignment when feature enabled.
* in order to configure the alignment settings when feature is enabled some CLI commands will be needed.
* add new MSP command for get/set arbitrary alignment settings.
* make it so the feature, cli commands, settings, msp commands, etc. can be conditionally compiled out.

hydra commented:
After other devs weighed-in on this in the slack channel, the idea is now as follows:

* store old and new alignment settings
* add 'custom' to list of old alignment settings
* avoid matrix computation unless 'custom' is used.
* add support for 'custom' settings to MSP.

 @mikeller please confirm if 'custom' mode needs to be able to be compiled out.

mikeller commented:
@hydra: Yes, this looks like a good way to do this to me.

hydra commented:
rebased on master, prior to rework.

hydra commented:
rebased on master again.

hydra commented:
After confirmation from @mikeller that the refactored approach is OK future work will continue on #8474.

This PR contains reference code that was released in the SPRacingH7EXTREME and SPRacingH7NANO firmware.
